Specify nodes' behavior to mitigate IL equivocation (#19)
soispoke/EIPs: EIPS/eip-FOCIL.md

  By default, ILs are built by selecting raw transactions from the public mempool, ordered by priority fees, up to the IL’s maximum size in bits (e.g., 8 KB per IL). Additional rules can be optionally applied to maximize censorship resistance, such as prioritizing valid transactions that have been pending in the mempool the longest.

#### Validators
- **`Slot N`, `t=0 to 9s`**: Validators receive ILs from the P2P network and store (1) all new ILs that pass the CL P2P validation rules, and any evidence of IL equivocation by committee members (i.e., if multiple ILs are received from the same committee member).

> - **`Slot N`, `t=9s`**: View freeze deadline. At this point, validators freeze their IL view and stop caching new ILs in memory.

Reviewer: I'd maybe just merge this point with the next. There's something specific that happens *at* the deadline, the whole behavior is what happens between 9s and 4s

- **`Slot N`, `t=9s` to `Slot N+1`, `t=4s`**: After the view freeze deadline, validators:
  1. Do not store new ILs received after the deadline.
  2. Continue forwarding ILs to peers following the CL P2P validation rules.
  3. Record any evidence of IL equivocation that occurs after the view freeze deadline.

After the attestation deadline of **`Slot N+1`, `t=4s`**, validators ignore any new ILs related to the previous slot's IL committee, and stop recording equivocation evidence for the previous slot's ILs.

#### Block Producer

> - **`Slot N`, `t=0 to 11s`**: The block producer (i.e., a proposer or a proposer builder pair) receive ILs from the P2P network, forwarding and caching those that pass the CL P2P validation rules. Optionally, an RPC endpoint can be added to allow the block producer to request missing ILs from its peers (e.g., by committee index at `t=10s`).

Reviewer: Proposer or proposer builder pair is a bit unclear to me. What is a proposer builder pair? If it's meant to be a local block builder (a proposer who also builds), this is confusing because then "just a proposer" would mean a proposer using mev-boost, and a proposer using mev-boost does not care about ILs at all since it does not build the payload. If instead a proposer builder pair is a proposer using mev-boost while a proposer is a local block builder, I think it's also confusing because in the "proposer builder pair" it is only actually the builder that deals with ILs. 

I would rephrase it to "a proposer doing local block building or an external builder", and perhaps change "block producer" just to builder, since here we are actually talking about the "payload producer", not the beacon block producer

Reviewer: Both for clarity purposes but also because it's actually a strength of this mechanism that it does not require (non self building) beacon block proposers to do anything

Author: agreed, made the following changes for clarity: 

```
#### Builder
- **`Slot N`, `t=0 to 11s`**: The builder (i.e., a proposer doing local block building or an external builder) receive ILs from the P2P network, forwarding and caching those that pass the CL P2P validation rules. Optionally, an RPC endpoint can be added to allow the builder to request missing ILs from its peers (e.g., by committee index at `t=10s`).

- **`Slot N`, `t=11s`**:
The builder freezes its view of ILs and asks the EL to update its execution payload by adding transactions from its view (the exact timings will be defined after running some tests/benchmarks).

#### Proposer
- **`Slot N+1`, `t=0s`**:
The proposer broadcasts its block with the up-to-date execution payload satisfying IL transactions over the P2P network.
```

  By default, ILs are built by selecting raw transactions from the public mempool, ordered by priority fees, up to the IL’s maximum size in bits (e.g., 8 KB per IL). Additional rules can be optionally applied to maximize censorship resistance, such as prioritizing valid transactions that have been pending in the mempool the longest.

#### Nodes
- **`Slot N`, `t=0 to 9s`**: Nodes receive ILs from the P2P network and store (1) all new ILs that pass the CL P2P validation rules, and any evidence of IL equivocation by committee members (i.e., if multiple ILs are received from the same committee member).

> - **`Slot N`, `t=9s`**: IL freeze deadline. At this point, nodes freeze their IL view and stop caching new ILs in memory.

Reviewer 2: We're using `IL freeze deadline` in the EIP but `VIEW_FREEZE_DEADLINE` in the consensus spec for the same purpose. Is it going to worth to use the same name to help readers understand?

Author: I'd keep IL freeze deadline here tbh, I feel like it's more intuitive (might be worth changing it in the specs though)

Reviewer: The reason I would call it view freeze deadline is that it might well end up being used for other things as well

Author: can it just be freeze deadline?

Author: changed in latest commit

The proposer broadcasts its block with the up-to-date execution payload satisfying IL transactions over the P2P network.

#### Attesters
- **`Slot N+1`, `t=4s`**:

> Attesters monitor the P2P network for the proposer’s block. Upon detecting it, they verify whether all transactions from their stored ILs are included in the proposer’s execution payload. The `Valid` function, based on the frozen view of the ILs from `t=9s` in the previous slot, checks if the execution payload satisfies IL validity conditions. This is done either by confirming that all transactions are present or by determining if any missing transactions are invalid when appended to the end of the payload. In such cases, attesters use the EL to perform nonce and balance checks to validate the missing transactions and check whether there is enough space in the block to include the transaction(s).

Reviewer: perhaps add "unless an equivocation from their sender has been detected"

> Attesters monitor the P2P network for the proposer’s block. Upon detecting it, they verify whether all transactions from their stored ILs are included in the proposer’s execution payload. The `Valid` function, based on the frozen view of the ILs from `t=9s` in the previous slot, checks if the execution payload satisfies IL validity conditions. This is done either by confirming that all transactions are present or by determining if any missing transactions are invalid when appended to the end of the payload. In such cases, attesters use the EL to perform nonce and balance checks to validate the missing transactions and check whether there is enough space in the block to include the transaction(s).

Reviewer: ```suggestion
Attesters monitor the P2P network for the proposer’s block. Upon detecting it, they verify whether all transactions from their stored ILs are included in the proposer’s execution payload, except for ILs whose sender has equivocated. The `Valid` function, based on the frozen view of the ILs from `t=9s` in the previous slot, checks if the execution payload satisfies IL validity conditions. This is done either by confirming that all transactions are present or by determining if any missing transactions are invalid when appended to the end of the payload. In such cases, attesters use the EL to perform nonce and balance checks to validate the missing transactions and check whether there is enough space in the block to include the transaction(s).
```

#### CL P2P Validation Rules

When validators receive ILs from the P2P network, they perform a series of validation checks before forwarding or caching them. These rules protect against Denial-of-Service (DoS) attacks by (1) limiting ILs' byte size and (2) restricting IL proposals to a small committee of IL committee members, thereby tightly bounding the main resource consumed by the propagation of ILs. Consumption of other relevant resources, such as verification time, is minimal because the only nontrivial check performed on IL propagation is signature verification. At this stage, there is no EL verification of the transactions within ILs. This means that ILs are allowed to contain any transactions—valid or invalid—since validators do not perform EL-side checks. This design choice is intended to avoid additional computational overhead.

1. The slot of the IL matches the current or previous slot.

> 2. The root of the IL committee referenced in the IL matches the expected IL committee root for the current or previous slot.

Reviewer: ```suggestion
2. The root of the IL committee referenced in the IL matches the expected IL committee root for its slot.
```

#### Fork choice changes

- Store ILs observed over gossip before the view freeze deadline.

> - If more than one IL is observed from the same IL committee member, remove all ILs from the member from the store.

Reviewer: ```suggestion
- If more than one IL is observed from the same IL committee member, mark the committee member as an equivocator and ignore any further ILs from them.
```

> - Fork choice head retrieval is based on the `Valid` function being satisfied by the EL.

Reviewer: ```suggestion
- The beacon block from the current slot is only attested to if the `Valid` function is satisfied, based on all stored ILs from non equivocators.
```
